File: backend/src/main/kotlin/mymoney/domain/model/Category.kt
package mymoney.domain.model

import kotlinx.datetime.Instant
import java.util.UUID

/**
 * Category (категория) — two-level hierarchy: parent categories and their
 * children. Categories are typed as income or expense. System categories are
 * pre-seeded per family; user categories are created by the user.
 *
 * System categories cannot be physically deleted, only archived
 * (см. § 25.1 Bible, closing open question № 5).
 */
data class Category(
    val id: UUID,
    val familyId: UUID,
    val parentCategoryId: UUID? = null,
    val name: String,
    val type: CategoryType,
    val isMandatory: Boolean = false,
    val isSystem: Boolean = false,
    val icon: String? = null,
    val color: String? = null,
    val isArchived: Boolean = false,
    val isDeleted: Boolean = false,
    val createdAt: Instant,
    val updatedAt: Instant,
)

File: mobile/lib/domain/model/category.dart
import 'enums.dart';

class Category {
  final String id;
  final String familyId;
  final String? parentCategoryId;
  final String name;
  final CategoryType type;
  final bool isMandatory;
  final bool isSystem;
  final String? icon;
  final String? color;
  final bool isArchived;
  final bool isDeleted;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Category({
    required this.id,
    required this.familyId,
    this.parentCategoryId,
    required this.name,
    required this.type,
    this.isMandatory = false,
    this.isSystem = false,
    this.icon,
    this.color,
    this.isArchived = false,
    this.isDeleted = false,
    required this.createdAt,
    required this.updatedAt,
  });

  /// `true`, если у категории есть родитель — то есть она подкатегория.
  bool get isSubcategory => parentCategoryId != null;

  Category copyWith({
    String? name,
    String? parentCategoryId,
    // Отдельный флаг, потому что `null` в parentCategoryId означает
    // «не менять», а не «сделать категорию корневой».
    bool clearParent = false,
    bool? isMandatory,
    String? icon,
    String? color,
    bool? isArchived,
    bool? isDeleted,
    DateTime? updatedAt,
  }) =>
      Category(
        id: id,
        familyId: familyId,
        parentCategoryId:
            clearParent ? null : (parentCategoryId ?? this.parentCategoryId),
        name: name ?? this.name,
        type: type,
        isMandatory: isMandatory ?? this.isMandatory,
        isSystem: isSystem,
        icon: icon ?? this.icon,
        color: color ?? this.color,
        isArchived: isArchived ?? this.isArchived,
        isDeleted: isDeleted ?? this.isDeleted,
        createdAt: createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );
}
